src/freenet/support/io/Fallocate.java

```diff
@@ -119,20 +127,22 @@ private static int getDescriptor(FileDescriptor descriptor) {
       field.setAccessible(true);
       return (int) field.get(descriptor);
     } catch (final Exception e) {
-      throw new UnsupportedOperationException("unsupported FileDescriptor implementation", e);
+      // File descriptor is not supported: fd is null and unavailable, return 0
+      return 0;
     }
   }
 
   private static void legacyFill(FileChannel fc, long newLength, long offset) throws IOException {
-    MersenneTwister mt = new MersenneTwister();
-    byte[] b = new byte[4096];
-    ByteBuffer bb = ByteBuffer.wrap(b);
-    while (offset < newLength) {
-      bb.rewind();
-      mt.nextBytes(b);
-      offset += fc.write(bb, offset);
-      if (offset % (1024 * 1024 * 1024L) == 0) {
-        mt = new MersenneTwister();
+    if (IS_WINDOWS) {
+      // Windows do not create sparse files by default, so just write a byte at the end.
```

Comment:
I don’t think we should assume that the user has whatever setting at the default value. Either verify that sparse files are disabled (which you would probably have to do on every access because it can change at any time), or don’t assume anything and always create files with the method that will allocate all of the space.

Reply:
It's not a setting for the user, only the application can change this setting as far as I know (https://learn.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-fsctl_set_sparse). The user can change an existing file into a sparse file via `fsutill` (https://learn.microsoft.com/zh-cn/windows-server/administration/windows-commands/fsutil-sparse), but not for new files.